### PR TITLE
Add preliminary evdev discovery

### PR DESCRIPTION
**Pull request by TheNeikos**

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
First a basic question: Why is this in a new module? Isn't it a compositor thing?

**Comment by TheNeikos** [comment]
Input is not something the compositor should handle. It should just compose (render the final image).

**Comment by matthiasbeyer** [comment]
Hm. @neithernut what are you thinking? The problem is that it differs from our inital planning of the modules, you know? That's why I ask.

**Comment by neithernut** [comment]
We already decided to have an input module, if you remember...

File: src/input/module.c [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        return -1;
+    }
+
+
+    while ((dir = readdir(d)) != NULL) {
+        if (!filter_name(dir->d_name)) {
+            continue;
+        }
+        char full_path[100] = "/dev/input/";
+        strcat(full_path, dir->d_name);
```
Comment (neithernut) [inline src/input/module.c]: would ust `strncat` here.

**Comment by matthiasbeyer** [comment]
Huh, right... but that was a long time ago, right? So, well... I'll review now.

File: src/input/module.c [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+#include "objects/set.h"
+
+static struct ws_logger_context log_ctx = { .prefix = "[Input] " };
+
+static struct ws_set devices;
+
+#define INPUT_PATH "/dev/input/"
+#define DEVICE_NAME "event"
+
+static void
```
Comment (matthiasbeyer) [inline src/input/module.c:54]: Overall function documentation missing.

File: src/input/module.c [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        event = (struct inotify_event*)
+            (((char*) event) + event->len + sizeof(*event));
+    }
+
+
+}
+
+static bool
+filter_name(
+    const char* name
```
Comment (matthiasbeyer) [inline src/input/module.c:94]: Overall function documentation missing as well as for this parameter.

File: src/input/module.c [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+) {
+    if (!(strcmp(".", name) && strcmp("..", name))) {
+        return false;
+    }
+    if (strncmp(DEVICE_NAME, name, sizeof(DEVICE_NAME))) {
```
Comment (matthiasbeyer) [inline src/input/module.c]: Do you really need an `strncmp()` call here? `strcmp()` should work fine, as `DEVICE_NAME` is null-terminated, isn't it?

Comment (matthiasbeyer) [inline src/input/module.c]: Also you can do a `return (0 == strcmp(...` here.

File: src/input/module.c [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        return false;
+    }
+    return true;
+}
+
+static int
```
Comment (matthiasbeyer) [inline src/input/module.c:112]: Function documentation missing.

File: src/input/module.c [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+find_initial_devices()
```
Comment (matthiasbeyer) [inline src/input/module.c]: `void` inside the parentheses missing.

File: src/input/module.c [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+{
```
Comment (matthiasbeyer) [inline src/input/module.c]: You can put the open bracket on the line above if you do not get any arguments here:

``` C
void
foo(void) {
```

File: src/input/module.c [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    DIR* d;
+    struct dirent* dir;
+
+    d = opendir(INPUT_PATH);
+
+    if (!d) {
+        ws_log(&log_ctx, LOG_CRIT, "Could not open input device director");
```
Comment (matthiasbeyer) [inline src/input/module.c]: "director"? You mean "directory" here, right?

File: src/input/module.c [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        return -1;
+    }
+
+
+    while ((dir = readdir(d)) != NULL) {
+        if (!filter_name(dir->d_name)) {
+            continue;
+        }
+        char full_path[100] = INPUT_PATH;
```
Comment (matthiasbeyer) [inline src/input/module.c]: Where comes this magic constant from? Please document it.

File: src/input/module.c [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    if (watchfd < 0) {
+        ws_log(&log_ctx, LOG_CRIT, "Could not watch input devices");
+        return watchfd;
+    }
+
+    struct ev_io* watcher = calloc(1, sizeof(*watcher));
+    ev_io_init(watcher, watch_devices, inotfd, EV_READ);
+
+    struct ev_loop* default_loop = ev_default_loop(EVFLAG_AUTO);
+    if (!default_loop) return -1;
```
Comment (matthiasbeyer) [inline src/input/module.c]: Brackets missing.

File: src/input/module.c (hunk 4 above)
Comment (TheNeikos) [inline src/input/module.c]: Nope, if S1 < S2 it will return a wrong value

**Comment by neithernut** [comment]
ACK.

**Comment by matthiasbeyer** [comment]
ACK.

**Pull request merged**